drivers/platform/aducm3029/irq.c

```diff
@@ -0,0 +1,302 @@
+ */
+int32_t irq_ctrl_init(struct irq_desc **desc,
+		      const struct irq_init_param *param)
+{
+	struct aducm_irq_desc		*aducm_desc;
+
+	if (!desc || !param || param->irq_id >= NB_INTERRUPT_CONTROLLERS ||
+	    initialized)
+		return FAILURE;
+
+	if (!(*desc = calloc(1, sizeof(**desc))))
+		return FAILURE;
+	if (!(aducm_desc = calloc(1, sizeof(*aducm_desc)))) {
+		free(*desc);
+		*desc = NULL;
+		return FAILURE;
+	}
+
+	(*desc)->extra = aducm_desc;
+	(*desc)->irq_id = param->irq_id;
+
+	adi_xint_Init(aducm_desc->irq_memory, ADI_XINT_MEMORY_SIZE);
+
+	initialized = 1;
+	return SUCCESS;
+}
+
+/**
+ * @brief Free the resources allocated by \ref irq_ctrl_init()
+ * @param desc - Interrupt controller descriptor.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_ctrl_remove(struct irq_desc *desc)
+{
+	if (!desc || !desc->extra || !initialized)
+		return FAILURE;
+
+	initialized = 0;
+	adi_xint_UnInit();
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Registers a generic IRQ handling function
+ * @param desc - Interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt
+ * @param irq_handler - Generic function to be registered. Will be called using
+ * as parameter the interrupt ID.
+ * @param dev_instance - Specify the trigger condition for the interrupt. To be
+ * one of the values from \ref irq_mode.
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_register(struct irq_desc *desc, uint32_t irq_id,
+		     void (*irq_handler)(void *data), void *dev_instance)
+{
+	struct aducm_irq_desc *aducm_desc;
+	if (!desc || !desc->extra || !initialized || !irq_handler ||
+	    irq_id >= NB_EXT_INTERRUPTS)
+		return FAILURE;
+
+	aducm_desc = desc->extra;
+	aducm_desc->irq_handler[irq_id] = irq_handler;
+	aducm_desc->mode[irq_id] = (enum irq_mode)dev_instance;
+
+	adi_xint_RegisterCallback(id_map_event[irq_id], internal_callback,
+				  aducm_desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Unregisters a generic IRQ handling function
+ * @param desc - Interrupt controller descriptor.
+ * @param irq_id - Id of the interrupt
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t irq_unregister(struct irq_desc *desc, uint32_t irq_id)
+{
+	if (!desc || !desc->extra || !initialized ||
+	    irq_id >= NB_EXT_INTERRUPTS)
+		return FAILURE;
+	(void)irq_id;
+
+	irq_source_disable(desc, irq_id);
+	adi_xint_RegisterCallback(id_map_event[desc->irq_id], NULL, 0);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Enable all previously enabled interrupts by \ref irq_source_enable().
+ * @param desc - Interrupt controller descriptor.
+ * @return \ref SUCCESS
+ */
+int32_t irq_global_enable(struct irq_desc *desc)
+{
+	struct aducm_irq_desc *aducm_desc;
+	if (!desc || !desc->extra || !initialized)
+		return FAILURE;
+
+	aducm_desc = desc->extra;
+	for (uint32_t i = 0; i < NB_EXT_INTERRUPTS; i++) {
```

Comment:
nitpick: I don't think this needs brackets.

Reply:
true